kubernetes/schema_pod_spec.go

```diff
@@ -210,7 +210,8 @@ func podSpecFields(isUpdatable, isComputed bool) map[string]*schema.Schema {
 				Schema: map[string]*schema.Schema{
 					"name": {
 						Type:         schema.TypeString,
-						Required:     true,
+						Optional:     true,
+						Default:      "linux",
```

Comment:
Since this attribute is optional, I'm wondering if we really need to set a default. Being optional, the API likely already tolerates missing values. I would follow the principle of least surprise here and leave the default value out.

Reply:
`spec.OS` is [optional](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3230C1-L3231) but `spec.OS.Name` is [required](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L3321-L3327).

If we keep `name` optional with the default value `linux` we will always send this value to the Kubernetes API and Pods will have it set to `linux` which does not match the default Kubernetes behavior.  The default one -- if not sent, keep it empty. That is why now we return [`nil`](https://github.com/hashicorp/terraform-provider-kubernetes/blob/main/kubernetes/structures_pod.go#L710-L712) if `os.name` is not set.

I agree with what Alex said, I would keep it as it is now.

Reply:
I think the test failures observed here are because of the replacement of the `expandOS` function. I would keep the one we have right now.

We might want to update it though.

```golang
func expandOS(l []interface{}) *v1.PodOS {
	if len(l) == 0 || l[0] == nil {
		return nil
	}

	in := l[0].(map[string]interface{})
	obj := &v1.PodOS{}

	if v, ok := in["name"].(string); ok {
		obj.Name = v1.OSName(v)
	}

    // If the above assertion fails and we return the PodOS structure with an empty Name field
    // then that will cause an error message from the Kubernetes API. The name cannot be empty.
    // We either need to return obj within the if statement and return nil here 
    // or introduce an error return if the above assertion fails.
    // We even can think whether we need assertion since the name would be a string
    // due to its type in the schema and it would be one of two values since we do validation.
	return obj
}
```

I left some comments in the code snippet.

I think the current expandOS might potentially look like this:

```golang
func expandOS(l []interface{}) *v1.PodOS {
	if len(l) == 0 || l[0] == nil {
		return nil
	}

	in := l[0].(map[string]interface{})

	return &v1.PodOS{
		Name: v1.OSName(in["name"].(string)),
	}
}
```